### PR TITLE
Implemented the SR-IOV Multi VF device pass-through for MultiVM's.

### PR DESCRIPTION
**Pull request by santwana**
Implemented the SR-IOV Multi VF device pass-through for MultiVM's.
User can now create the desired no.of VM's along with the specified VF's(single or multiple),
to be pass-throughe'd to single/mutiple VM's.

Signed-off-by: Santwana Samantray <santwana@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by santwana** [comment]
[job.log](https://github.com/autotest/tp-libvirt/files/2256041/job.log)

File: libvirt/tests/src/libvirt_pci_passthrough_multi.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+            for val in virt_functions:
+                pci_dev = utils_test.libvirt.pci_label_from_address(val,
+                                                                    radix=16)
+                vf1.append(pci_dev)
+        # Check if sufficient VF's are available for pass-through to the guests.
+        if len(vf1) >= (int(no_vf) * len(vms)):
+            i = 0
+            for vm in vms:
+                for _ in range(int(no_vf)):
+                    pci_xml = NodedevXML.new_from_dumpxml(vf1[i])
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_multi.py]:
```python
for each in range(int(no_vf)):
    pci_xml = NodedevXML.new_from_dumpxml(vf1[each])
```

File: libvirt/tests/src/libvirt_pci_passthrough_multi.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    try:
+        for vm in vms:
+            bus_info = []
+            vm.start()
+            session = vm.wait_for_login()
+            # The Network configuration is generic irrespective of PF or SRIOV VF
+            output = session.cmd_output("lspci -nn")
+            nic_list_after = output.splitlines()
+            net_ip = netaddr.IPAddress(net_ip)
+            if nic_list_after == nic_list_before:
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_multi.py]: It is better to sort the list for asserting,

File: libvirt/tests/src/libvirt_pci_passthrough_multi.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                test.fail("passthrough Adapter not found in guest.")
+            else:
+                logging.debug("Adapter passthorughed to guest successfully")
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_multi.py]: typo for passthrough

File: libvirt/tests/src/libvirt_pci_passthrough_multi.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    # Check the parameters from configuration file.
+    if pci_dev.count("ENTER"):
+        test.cancel("Please enter your device name for test.")
+    if (net_ip.count("ENTER") or server_ip.count("ENTER") or netmask.count("ENTER")):
+        test.cancel("Please enter the ips and netmask for NIC "
+                    "test in config file")
+    for vm in vms:
+        vmxml = VMXML.new_from_inactive_dumpxml(vm.name)
+        backup_xml = vmxml.copy()
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_multi.py]: backup_xml for each vm should be saved in a dict, else each iteration will overwrite backup_xml

File: libvirt/tests/src/libvirt_pci_passthrough_multi.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                    i += 1
+        else:
+            test.cancel("Not enough VF's present for the pass-through")
+    else:
+        for val in pci_devs:
+            val = val.replace(".", "_")
+            pci_xml = NodedevXML.new_from_dumpxml(val)
+            pci_address = pci_xml.cap.get_address_dict()
+            vmxml.add_hostdev(pci_address)
+            vmxml.sync()
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_multi.py]: In multivm env, to which vmxml we are adding this hostdev ?

File: libvirt/tests/src/libvirt_pci_passthrough_multi.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                                                 session=session)
+                logging.info(o_ping)
+                if s_ping != 0:
+                    err_msg = "Ping test fails, error info: '%s'"
+                    logging.debug(err_msg % o_ping)
+                # Each interface should have unique IP
+                net_ip = net_ip + 1
+
+    finally:
+        backup_xml.sync()
```
Comment (balamuruhans) [inline libvirt/tests/src/libvirt_pci_passthrough_multi.py]: sync() the xml for all the vms

**Comment by santwana** [comment]
depends on: autotest/tp-libvirt#2179

**Comment by santwana** [comment]
@balamuruhans I have addressed the review changes and also rebased the base file to existing (https://github.com/autotest/tp-libvirt/pull/2179/). Kindly review. Thanks.

**Comment by sathnaga** [comment]
@santwana needs rebase.

**Comment by dzhengfy** [comment]
After confirmation with the IBM PPC FVT lead that the author does not work on this any more and the pr is not in use by them. So I close it.

**Pull request closed without merging**